### DISCUSSION TITLE
Passing variables back and forth between Python and AHK.

### DISCUSSION THREAD

**Discussion by burque505** [spyoungtech/ahk#341] [Q&A] [answered]
This is a fantastic project, really enjoying it. I'm having trouble, though, establishing a workflow for passing variables back and forth between Python and AHK.  I can do it by reading and writing text files, which works but seems inefficient and perhaps not very scalable.
At the moment I need to get the text from a Tkinter Text widget into AHK as part of a script to run with ahk.run_script(my_script_here). 

For example, I have a Text widget, 't1'.  Examining it with WindowSpy tells me the widget is TkChild2:

```
success_script='''\
outstuff := ControlGetText("TkChild2", "ahk_class TkTopLevel")
MsgBox(type(outstuff), "My Message")
MsgBox(outstuff, "Epic Fail")
'''
```
The first MsgBox shows the type of 'outstuff' is String, but takes approximately four (4) seconds to show any output.
The last MsgBox shows the title, but the body is empty.

Might I trouble someone for code that a) passes a Python string to a script like the one above, and b) passing an AHK variable to Python?

I would also very much like to see people's examples of working with both the Sync and Async APIs.

Thanks very much!
Best regards,
burque505

**Comment by spyoungtech** [comment]
The most robust approach that covers a lot of use cases would be to write [an extension](https://ahk.readthedocs.io/en/latest/extending.html).

For very simple use cases, if your AutoHotkey script writes to stdout, that will be returned from `run_script` as a string.

For example:

```python
# https://github.com/spyoungtech/ahk/blob/46bce5c3e19aaa9fe49f1a90816a80a96a39a2c2/ahk/_utils.py#L137-L146
script = '''\
#NoTrayIcon
version := Format("{}", A_AhkVersion)
filename := "*"
encoding := "UTF-8"
mode := "w"
stdout := FileOpen(filename, mode, encoding)
stdout.Write(version)
stdout.Read(0)
'''
result = ahk.run_script(script)
print(result)
```

**Reply by burque505** [reply]
Thanks, @spyoungtech, that's very helpful. I'm a ways away from trying to write an extension at the moment, though.
The reason I posted was primarily to see if I can get the following code to a) get the text of a Tkinter 'Text' widget and b) pass the variable containing that text to an AHK script (e.g. 'success_script' in the code below) and show that in a MsgBox. I realize my import statements are not correct yet.

```
import ttkbootstrap as ttk
import tkinter as tk
from ttkbootstrap.constants import *
from ahk import AHK, MsgBoxButtons, Control
import ahk

#print(dir(AHK))


success_script='''\
outstuff := ControlGetText("TkChild2", "ahk_class TkTopLevel")
MsgBox(type(outstuff), "My Message")
MsgBox(outstuff, "Epic Fail")
'''

def on_button_click():
   """
   Handles the event triggered when a button is clicked.
   
   Parameters:
   None
   
   Returns:
   None
   """
   print("Button clicked!")

def on_secondary_button_click():
   """
   Handles the event triggered when the secondary button is clicked.
   
   Parameters:
   None
   
   Returns:
   None
   """
   ahk.msg_box(text='Do you like message boxes?', title='My Title', buttons=MsgBoxButtons.YES_NO)
   
def on_success_button_click():
   text_to_print = t1.get("1.0", "end-1c")
   print(text_to_print)
   
def on_info_button_click():
   ahk.run_script(success_script)
   text_to_print = str(t1.get("1.0", "end-1c"))
   print(text_to_print)


ahk=AHK()   
message = "Hello from the secondary button!"


   
ttk.utility.enable_high_dpi_awareness()
root = ttk.Window()
root.title("Lambda callback example")

b1 = ttk.Button(root, text='primary', command=on_button_click, bootstyle=PRIMARY)
b1.pack(side=LEFT, padx=5, pady=5)

b2 = ttk.Button(root, text='secondary', command=on_secondary_button_click, bootstyle=SECONDARY)
#b2 = ttk.Button(root, text='secondary', command=lambda: on_secondary_button_click(), bootstyle=SECONDARY)
b2.pack(side=LEFT, padx=5, pady=5)

b3 = ttk.Button(root, text='success', command=lambda: on_success_button_click(), bootstyle=SUCCESS)
b3.pack(side=LEFT, padx=5, pady=5)

b4 = ttk.Button(root, text='info', command=on_info_button_click, bootstyle=INFO)
b4.pack(side=LEFT, padx=5, pady=5)

b5 = ttk.Button(root, text='warning', bootstyle=WARNING)
b5.pack(side=LEFT, padx=5, pady=5)

b6 = ttk.Button(root, text='danger', bootstyle=DANGER)
b6.pack(side=LEFT, padx=5, pady=5)

b7 = ttk.Button(root, text='light', bootstyle=LIGHT)
b7.pack(side=LEFT, padx=5, pady=5)

b8 = ttk.Button(root, text='dark', bootstyle=DARK)
b8.pack(side=LEFT, padx=5, pady=5)

t1 = tk.Text(master=root)
t1.pack(side=LEFT, padx=5, pady=5)

t1.insert(1.0, success_script)


root.mainloop()
```
My problem appears to center around the ControlGetText statement in 'success_script'. It's obviously getting a string, but apparently not the string that's in 't1' (aka 'TkChild2' here0.

Best regards,
burque505

**Reply by spyoungtech** [reply, marked as answer]
I think this is a general problem with how Tkinter renders text. I've never had much luck getting text from tkinter windows at all. For example, using AutoHotkey's window spy, you won't see the text, either.

**Comment by burque505** [comment]
I'm sure you're right! That's my experience with WindowSpy also. Very curious. I can get the Hwnd of the tkinter Text widget, but not its contents (in AHK, that is).
